src/index.js

```diff
@@ -16,102 +17,141 @@ function makeDnsPromise(host) {
   var endTime = 599900; /* Maximum execution: 100ms short of 5 minutes */
   if (config.timeoutWindow < 1) {
     return undefined;
   }
 
-  if (config.timeoutWindow > 0 && context.getRemainingTimeInMillis) {
+  if (config.timeoutWindow > 0 && modifiedContext.getRemainingTimeInMillis) {
     endTime = Math.max(
       0,
-      context.getRemainingTimeInMillis() - config.timeoutWindow
+      modifiedContext.getRemainingTimeInMillis() - config.timeoutWindow
     );
   }
 
   return setTimeout(() => {
-    report.send(new Error('Timeout Exceeded.'), function noop() {});
+    sendReport(new Error('Timeout Exceeded.'));
   }, endTime);
 }
 
-module.exports = options => {
-  const fn = func => {
-    var dnsPromise = undefined;
-    fn.metricsQueue = [];
-    const config = setConfig(options);
-
-    if (!config.clientId) {
-      // No-op if user doesn't set an IOpipe token.
-      return func;
+class IOpipeWrapperClass {
+  constructor(userFunc, originalEvent, originalContext, originalCallback) {
+    if (!globals.COLDSTART) {
+      /* Get an updated DNS record. */
+      dnsPromise = makeDnsPromise(config.host);
     }
 
-    /* resolve DNS early on coldstarts */
-    dnsPromise = makeDnsPromise(config.host);
-
-    return function IOpipeWrapper() {
-      fn.metricsQueue = [];
-      const args = [].slice.call(arguments);
-
-      if (!globals.COLDSTART) {
-        /* Get an updated DNS record. */
-        dnsPromise = makeDnsPromise(config.host);
-      }
+    this.metrics = [];
+    this.originalContext = originalContext;
+    this.originalCallback = originalCallback;
+    this.report = new Report(
+      config,
+      this.originalContext,
+      process.hrtime(),
+      this.metrics,
+      dnsPromise
+    );
 
-      const startTime = process.hrtime();
-      const report = new Report(
-        config,
-        args[1],
-        startTime,
-        fn.metricsQueue,
-        dnsPromise
+    // preserve original functions via a property name change
+    ['succeed', 'fail', 'done'].forEach(method => {
+      Object.defineProperty(
+        this.originalContext,
+        `original_${method}`,
+        Object.getOwnPropertyDescriptor(this.originalContext, method)
       );
+    });
 
-      const timeout = setupTimeoutCapture(config, report, args[1]);
-
-      const callback = (err, cb) => {
-        if (timeout) {
-          clearTimeout(timeout);
-        }
-        report.send(err, cb);
-      };
-
-      /* Mangle arguments, wrapping callbacks. */
-      args[1] = Context(callback, args[1]);
-      args[2] = Callback(callback, args[2]);
-
-      try {
-        return func.apply(this, args);
-      } catch (err) {
-        clearTimeout(timeout);
-        report.send(err, function noop() {});
-        return undefined;
+    // assign modified methods and objects here
+    this.modifiedContext = Object.assign(this.originalContext, {
+      // need to use .bind, otherwise, the this ref inside of each fn is NOT IOpipeWrapperClass
```

Comment:
One issue that I found while working on this refactor was that we were selectively calling out `getRemainingTimeInMillis` as a prop that we wanted to preserve for aws' sake. The issue could be that aws decides to add a new prop, then we have to ship new code to support it. So I inverted the approach to only _add_ the items that we wanted to, while preserving the original object (and object refs). This is done by effectively renaming the old `.success`, `.fail` etc to `original_success` etc, then simply `Object.assign`ing the new stuff that we want to add. You can see a test of this in action here:
https://github.com/coreylight/iopipe/blob/7d95343a5c8554e8bdd741c3b6f1d8665220e5c1/src/index.test.js#L94
I don't have a specific test for `getRemainingTimeInMillis` but I could add one!